crates/sui-json-rpc/src/name_service.rs

```diff
@@ -154,12 +160,59 @@ impl FromStr for Domain {
             return Err(DomainParseError::LabelsEmpty);
         }
 
+        add_tld_if_missing(&mut labels);
+
         let labels = labels.into_iter().map(ToOwned::to_owned).collect();
 
         Ok(Domain { labels })
     }
 }
 
+// Parses a separator from the domain string input.
+// E.g.  `example.sui` -> `.` | example@sui -> `@` | `example*sui` -> `*`
+fn separator(s: &str) -> Result<char, DomainParseError> {
+    let mut domain_separator: Option<char> = None;
+
+    for separator in ACCEPTED_SEPARATORS.iter() {
```

Comment:
curious what's the discussion to support @ and * as well; before this, for example I want to claim a domain of `abcd`, I only need to own `abcd.sui`, but now I will have to own `abcd*sui` and `abcd@sui`?

Reply:
`abcd.sui` is a single name on-chain. Resolution will now work with all these different variations.

Querying for `@abcd` or `abcd@sui` or `abcd*sui` will resolve for the on-chain `abcd.sui`.

Querying for `example@abcd` will resolve for the on-chain `example.abcd.sui`.

Reply:
im also curious as to why we'd want to handle "@" and "*" instead of only allowing "."

Reply:
Also slightly confused about this aspect of the design, is there somewhere we can go to read and learn more about it @manolisliolios?

I can perhaps see the product value of `@abcd`, but then `example@abcd` starts to look like an email address, which leans people towards `example@abcd.sui` which is then not allowed because it contains multiple separators, meanwhile I don't really have any context clues for what `*` is intended for.

The `example@abcd.sui` vs `@abcd` case also highlights that in conventional domain names, the `@` part is not part of a domain, it's usually part of the [`authority` part of a URI](https://en.wikipedia.org/wiki/Uniform_Resource_Identifier#Syntax), which may or may not include a domain name.

That all leads me to the same question that @wlmyng had, of why we need to offer these options in the RPC layer.  Domain names are resource identifiers, and it's helpful to make them canonical for that reason.  If a product (e.g. the wallet) wants to represent them in some contracted way (e.g. treating `@foo` as `foo.sui`) then should we bake that into the domain name format, or let the product make the transformation? What if a hypothetical suitter app wanted to treat `@foo` as `foo.suitter.sui`?

Reply:
There will be different (but strictly defined) variations on the visualisation of names, and supporting all these different inputs makes RPC resolution much easier across apps, while also giving the chance for more per-product customizations.

For different products, it's not really recommended to divert from the RPC-supported formats, because of the discoverability utilities that come with a valid representation of a name (sending assets, look up on explorers etc).

With this, any explorer/wallet can support `@name`, `test@name`, `test*name`, `test.name`, `test.name.sui` (and so on) resolution out of the box.

Lastly, we'll also be moving to `@` as our base formatting (with omitted TLD), and we are aware that `*` will also be heavily used.

Reply:
Adding a note here.
There was actually an on-going discussion about this.

It turns out that we won't omit TLDs, and the `@` is a special format that replaces the `.sui` version (only on the SLD level)
So `@test` -> `test.sui`
and `test@test` -> `test.test.sui`
but `inner.test@test` -> `inner.test.test.sui`

We can use either `.` or `*` as the overall separator but not both.